### DISCUSSION TITLE
Run K3s cli command from a pod?

### DISCUSSION THREAD

**Discussion by TheOnlyWei** [k3s-io/k3s#10727] [Q&A]
Is it possible to run `k3s` cli commands from a pod? For example, if I want to deploy a pod to run a `service.key` rotation periodically using the k3s cli command using:
```
k3s certificate rotate-ca --path=/opt/k3s/server
```
The best way I found to do this currently is to mount the entire server node file system into a pod, use the server node `hostNetwork`, `hostPID`, and a privileged `serviceAccountName` to `chroot` the mounted server node file system. Then I can run the `k3s` command.

However, taking a look at the k3s CLI code, it seems to make HTTP requests, at least for the `k3s certificate rotate-ca` command:
https://github.com/k3s-io/k3s/blob/876d54cf49ad1fc365141fb4039e09b70871638d/pkg/cli/cert/cert.go#L269

Is there a way to make such an HTTP request from a pod in the server node's cluster, or even just run the `k3s` command itself, or is this not a supported use case? Although, even granted that it is possible, I would still need to run some related bash commands remotely for the full rotation and so I would likely need to use the chroot command regardless. But it would be nice to know if the HTTP request method is somehow possible or if there is a way to run `k3s` command from a pod.

**Comment by brandond** [comment]
You could play with using the `rancher/k3s` image? You'd still probably need to mount some things from the host filesystem though.

You could probably look at what the k3s-upgrade image and k3s upgrade system-upgrade-controller plans are doing, as a source of inspiration.

**Reply by TheOnlyWei** [reply]
Even if it is possible (which might be a vulnerability anyway that has to be patched), is it recommended to for users to work around `k3s certificate rotate-ca` and do the HTTP call from a pod avoid without having to use the `hostNetwork`, `hostPID`, and mounting the host's root context? I am trying to minimize the level of privileges that a pod requires to do a service account signing key rotation. If we can be certain that mounting specific files and then making an HTTP call will work, then that is a smaller attack surface for a pod than one that has access to the `hostNetwork`, `hostPID`, and the host's root context. Then again, if it does work, then anyone can just recreate those files in any isolated context of any pod and do key rotations.
